internal/runtime/local/run.go

```diff
@@ -0,0 +1,127 @@
+
+package local
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/inspektor-gadget/inspektor-gadget/internal/runtime"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+func (r *Runtime) runGadget(runner runtime.Runner, gadget gadgets.GadgetInstantiate, operatorPerGadgetParamCollection params.Collection) error {
+	log := runner.Logger()
+
+	// Create gadget instance
+	gadgetInstance, err := gadget.NewInstance(runner)
+	if err != nil {
+		return fmt.Errorf("instantiate gadget: %w", err)
+	}
+
+	// Deferring getting results and closing to make sure operators got their chance to clean up properly beforehand
+	defer func() {
+		if closer, ok := gadgetInstance.(gadgets.CloseGadget); ok {
+			log.Debugf("calling gadget.Close()")
+			closer.Close()
+		}
+		if results, ok := gadgetInstance.(gadgets.GadgetResult); ok {
+			res, err := results.Result()
+			log.Debugf("setting result")
+			runner.SetResult(res, err)
+		}
+	}()
+
+	// Install operators
+	err = runner.Operators().PreGadgetRun(runner, gadgetInstance, operatorPerGadgetParamCollection)
+	if err != nil {
+		return fmt.Errorf("starting operators: %w", err)
+	}
+	defer runner.Operators().PostGadgetRun()
+	log.Debugf("found %d operators", len(runner.Operators()))
+
+	if gadget.Type() == gadgets.TypeTraceIntervals {
+		// Enable interval pushes
+		log.Debugf("enabling snapshots")
+		runner.Parser().EnableSnapshots(runner.Context(), time.Second, 2)
+	}
+
+	// Set event handler
+	if setter, ok := gadgetInstance.(gadgets.EventHandlerSetter); ok {
+		log.Debugf("set event handler")
+		switch gadget.Type() {
+		default:
+			setter.SetEventHandler(runner.Parser().EventHandlerFunc(runner.Operators().Enrich))
+			// setter.SetEventHandler(runner.Parser().EventHandlerFuncNew(runner.Operators().Enricher))
+		}
+	}
+
+	// Set event handler for array results
+	if setter, ok := gadgetInstance.(gadgets.EventHandlerArraySetter); ok {
+		log.Debugf("set event handler for arrays")
+		switch gadget.Type() {
+		default:
+			setter.SetEventHandlerArray(runner.Parser().EventHandlerFuncArray(runner.Operators().Enrich))
+			// setter.SetEventHandlerArray(runner.Parser().EventHandlerFuncArrayNew(runner.Operators().Enricher))
+		case gadgets.TypeTraceIntervals:
+			setter.SetEventHandlerArray(runner.Parser().EventHandlerFuncSnapshot("main", runner.Operators().Enrich)) // TODO: "main" is the node
+			// setter.SetEventHandlerArray(runner.Parser().EventHandlerFuncSnapshotNew("main", runner.Operators().Enricher)) // TODO: "main" is the node
+		}
+	}
+
+	// Set event handler (currently only used by profile/cpu)
+	if setter, ok := gadgetInstance.(gadgets.EventEnricherSetter); ok {
+		log.Debugf("set event enricher")
+		setter.SetEventEnricher(runner.Operators().Enrich)
+		// setter.SetEventEnricher(runner.Operators().Enricher(func(a any) error {
+		// 	return nil
+		// }))
+	}
+
+	if startstop, ok := gadgetInstance.(gadgets.StartStopAltGadget); ok {
```

Comment:
Because some gadgets already had a Start() or Stop() method with a different signature - and as I wanted everything to be backwards compatible / not do more changes to gadgets than needed, I added them temporarily.

Reply:
I'm sorry, I still don't understand the need for both. Is it only to remember that this [`Stop`](https://github.com/inspektor-gadget/inspektor-gadget/blob/michael/gadget-registry-local/pkg/gadgets/profile/block-io/tracer/tracer.go#L171) needs to be modified to match `StartStopGadget`?

Reply:
> I'm sorry, I still don't understand the need for both. Is it only to remember that this [Stop](https://github.com/inspektor-gadget/inspektor-gadget/blob/michael/gadget-registry-local/pkg/gadgets/profile/block-io/tracer/tracer.go#L171) needs to be modified to match `StartStopGadget`?

Yes, after refactoring everything, they all should have the same signature for Start/Stop and we'd deprecate/remove `StartStopAltGadget`.